lib/tasks/version_scheme_counter.rake

```diff
@@ -0,0 +1,108 @@
+#   semver, but this may turn out to be the wrong choice later on.
+
+module VersionSchemeDetection
+  REGEXES = {
+    MAVEN: /^(0|([1-9]\d*))(\.[1-9]\d*){1,2}(-(((alpha|beta|rc)-)?\d+|SNAPSHOT))?$/,
+    # PEP440 Regex copied from https://www.python.org/dev/peps/pep-0440/#appendix-b-parsing-version-strings-with-regular-expressions
+    PEP440: /^([1-9][0-9]*!)?(0|[1-9][0-9]*)(\.(0|[1-9][0-9]*))*((a|b|rc)(0|[1-9][0-9]*))?(\.post(0|[1-9][0-9]*))?(\.dev(0|[1-9][0-9]*))?$/
+  }
+
+  # validators return truthy/falsy
+  VALIDATORS = {
+    MAVEN: ->(version) { version =~ REGEXES[:MAVEN] },
+    PEP440: ->(version) { version =~ REGEXES[:PEP440] },
+    SEMVER: ->(version) { SemanticRange.valid(version).present? },
+    CALVER: lambda do |version|
+      first_part = version.split(".")[0]
+      # Assume most if not all Calver versions were released after 1000 AD
+      # Also assume if they are using YY as the first part that it is after 2000 AD
+      first_part.length >= 4 || (first_part.length == 2 && first_part.to_i > 0)
+    end
+  }
+end
+
+tallies = { semver: 0, pep440: 0, maven: 0, calver: 0, unknown: 0, no_versions: 0 }
+
+namespace :version do
+  desc 'Tests a sampling of project\'s versions to count occurrences of different versioning schemes'
+  task scheme_counter: :environment do
+    # projects = Project.order(Arel.sql("RANDOM()")).limit(100)
+
+    global_tallies = tallies.clone.merge({
+      unknown_versions: []
+    })
+
+    Project.find_each(batch_size: 1000) do |project|
+      local_tallies = tallies.clone
+
+      if !project.versions_count?
+        global_tallies[:no_versions] += 1
+        next
+      end
+
+      project.versions.each do |version|
+        version_number = version.number
+        matchable = false
+        [
+          :semver,
+          :pep440,
+          :maven,
+          :calver,
+        ].each do |scheme|
+          if VersionSchemeDetection::VALIDATORS[scheme.upcase].call(version_number)
+            local_tallies[scheme] += 1
+            matchable = true
+          end
+        end
+
+        if !matchable
+          local_tallies[:unknown] += 1
+        end
+      end
+
+      max_count = local_tallies.values.max
+      maxes = local_tallies.select{|_, tally| tally == max_count}.keys
+
+      if maxes.length == 1
+        detected_scheme = maxes[0]
+      elsif maxes.include?(:unknown)
+        global_tallies[:unknown_versions].push([project.platform, project.name, project.versions.pluck(:number)])
+        detected_scheme = :unknown
+      elsif maxes.include?(:semver)
+        detected_scheme = :semver
+      elsif maxes.include?(:calver)
+        detected_scheme = :calver
+      elsif maxes.include?(:maven) && project.platform.downcase == "maven"
+        detected_scheme = :maven
+      elsif maxes.include?(:pep440) && project.platform.downcase == "pypi"
+        detected_scheme = :pep440
```

Comment:
Need to determine a fall-through tie-breaker

Reply:
Just to 2x here too, the ordering basically determines the priority that we give a certain scheme match, in case there are 2+ schemes that match the most?

Reply:
yes, that's correct